### ISSUE TITLE
Is BN updated for the whole 15 batches?

### ISSUE THREAD

**Issue by NovaMind-Z** [google-research/fixmatch#20]
Hi,
    Thanks for publishing your brilliant work! Recently I'm trying to reproducing this code by pytorch and I found some differences between fixmatch and mixmatch in the following lines: 
![image](https://user-images.githubusercontent.com/42997811/77843857-b6f88580-71d3-11ea-856c-9d9be21851c3.png)

1.  The interleave function in fixmatch doesn't make any change to the input x, so the interleave function in fixmatch is different with that in mixmatch, am Icorrect?
2.  In both your  issue and code of mixmatch, you mention that you only update BN mean and variance using the first batch(labeled inputs), but in this fixmatch, it seems you update BN using the whole 15 batches inputs(1 labeled and 14 unlabeled) , do I understand it correctly? If so, why use all the batches?

Thanks in advance for your reply

**Comment by david-berthelot** [comment]
1. Yes.
2. Yes.

There's no particular reason, we just wanted to keep the code as simple as possible for FixMatch.

**Comment by NovaMind-Z** [comment]
> 1. Yes.
> 2. Yes.
> 
> There's no particular reason, we just wanted to keep the code as simple as possible for FixMatch.

Thanks, and I think FixMatch is really simple and efficient.

**Comment by vfdev-5** [comment]
@NovaMind-Z thanks for asking about interleave op (me too, I'm trying to port the code to PyTorch).

@david-berthelot thanks the answer, however I still do not get why do we need to interleave. Could you please detail it ?

**Comment by david-berthelot** [comment]
We need to interleave due to multiple-GPU batch norm issues. Let's say we have to GPUs, and our batch is comprised of labeled (L) and unlabeled (U) images. Let's use a batch size of 2 for making easier visually in my following example. 

- Without interleaving, we have a batch `LLUUUUUU...U` (there are 14 U). When the batch is split to be passed to both GPUs, we'll have two batches `LLUUUUUU` and `UUUUUUUU`. Note that all labeled examples ended up in batch1 sent to GPU1. The problem here is that batch norm will be computed per batch and the moments will lack consistency between batches.

- With interleaving, by contrast, the two batches will be `LUUUUUUU` and `LUUUUUUU`. As you can notice the batches have the same distribution of labeled and unlabeled samples and will therefore have more consistent moments.

Hope this helps.

-

**Comment by vfdev-5** [comment]
@david-berthelot yes, now I understand better. Thank you very much for this explanation !

**Comment by haohang96** [comment]
@NovaMind-Z A minor questions for your answer: "In both your issue and code of mixmatch, you mention that you only update BN mean and variance using the first batch(labeled inputs), ". 

In mixmatch, the first batch, i.e. batch[0] in the code, is data after interleaved, so batch[0] is consisted of both labeled data and unlabeled data. So I think the first batch not means labeled inputs actually. 

If I was wrong, Please inform me. Thanks very much!

**Comment by JoyHuYY1412** [comment]
> We need to interleave due to multiple-GPU batch norm issues. Let's say we have to GPUs, and our batch is comprised of labeled (L) and unlabeled (U) images. Let's use a batch size of 2 for making easier visually in my following example.
> 
> * Without interleaving, we have a batch `LLUUUUUU...U` (there are 14 U). When the batch is split to be passed to both GPUs, we'll have two batches `LLUUUUUU` and `UUUUUUUU`. Note that all labeled examples ended up in batch1 sent to GPU1. The problem here is that batch norm will be computed per batch and the moments will lack consistency between batches.
> * With interleaving, by contrast, the two batches will be `LUUUUUUU` and `LUUUUUUU`. As you can notice the batches have the same distribution of labeled and unlabeled samples and will therefore have more consistent moments.
> 
> Hope this helps.

@david-berthelot 
Hi, thank you for your explanation. However, I am still confused why the labeled data need to be split equally to compute BN. Could you explain more about 

>  the moments will lack consistency between batches

when we compute BN with two batches `LLUUUUUU` and `UUUUUUUU`? How about the case `LLLLLLLL` and `UUUUUUUU`? Thank you.

**Comment by kgarg8** [comment]
Won't it look like `LULULULU` and `LULULULU`? Works, right?
